zerver/lib/push_registration.py

```diff
@@ -0,0 +1,94 @@
+import logging
+from typing import Any
+
+from django.conf import settings
+
+from zerver.lib.exceptions import (
+    InvalidBouncerPublicKeyError,
+    InvalidEncryptedPushRegistrationError,
+    RequestExpiredError,
+)
+from zerver.lib.remote_server import send_to_push_bouncer
+from zerver.models import PushDevice
+from zerver.models.users import UserProfile, get_user_profile_by_id
+from zerver.tornado.django_api import send_event_on_commit
+
+if settings.ZILENCER_ENABLED:
+    from zilencer.views import do_register_remote_push_device
+
+logger = logging.getLogger(__name__)
+
+
+def handle_registration_to_bouncer_failure(
+    user_profile: UserProfile, push_account_id: int, error_code: str
+) -> None:
+    PushDevice.objects.filter(user=user_profile, push_account_id=push_account_id).update(
+        error_code=error_code
+    )
+    event = dict(
+        type="push_device",
+        op="update",
+        push_account_id=str(push_account_id),
+        status="failed",
+        error_code=error_code,
+    )
+    send_event_on_commit(user_profile.realm, event, [user_profile.id])
+
+    if error_code == PushDevice.ErrorCode.REQUEST_EXPIRED_ERROR:
+        logging.error(
+            "Push device registration for user_id=%s, push_account_id=%s timedout.",
+            user_profile.id,
+            push_account_id,
+        )
+
+
+def handle_register_push_device_to_bouncer(event: dict[str, Any]) -> None:
```

Comment:
This `event` format would benefit from being a `TypedDict` so you get typing here that's shared with the code that generates and enqueues these events, rather than just declaring types in the liens below.

Reply:
`RegisterPushDeviceToBouncerPayload` payload suggests it's a network payload. It is in the sense that it goes through queues, but it does not get sent to the bouncer (`user_profile_id`  is secret from the bouncer).